File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import heapq
import sys
input = sys.stdin.readline


def union(x, y):
    x = find(x)
    y = find(y)

    # x와 y를 합칠 수 없음 (사이클 발생)
    if x == y:
        return False

    p[x] = y
    return True


def find(x):
    if p[x] != x:
        p[x] = find(p[x])
    return p[x]


def kruskal():
    ans = 0     # MST의 가중치 합
    cnt = 0     # 선택된 간선의 개수

    while cnt < V - 1:
        # 최소 비용 간선 선택하기
        e, v1, v2 = heapq.heappop(h)

        # 사이클 발생하는지 확인하기
        if union(v1, v2):
            ans += e
            cnt += 1

    return ans


V, E = map(int, input().split())

visited = [False] * (V + 1)
p = [i for i in range(V + 1)]  # 루트 노드
h = []  # 최소 힙

for _ in range(E):
    a, b, c = map(int, input().split())
    heapq.heappush(h, [c, a, b])

print(kruskal())

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
	
	static int[] p;
	
	static class Edge implements Comparable<Edge> {
		
		int v1;
		int v2;
		int weight;
		
		public Edge(int v1, int v2, int weight) {
			this.v1 = v1;
			this.v2 = v2;
			this.weight = weight;
		}

		@Override
		public int compareTo(Edge edge) {
			return Integer.compare(this.weight, edge.weight);
		}
	}
	
	public static void main(String[] args) throws Exception {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		PriorityQueue<Edge> pq = new PriorityQueue<>();
		
		int V = Integer.parseInt(st.nextToken());
		int E = Integer.parseInt(st.nextToken());
		
		p = new int[V + 1];
		for (int i = 1; i < V + 1; i++) {
			p[i] = i;
		}
		
		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int A = Integer.parseInt(st.nextToken());
			int B = Integer.parseInt(st.nextToken());
			int C = Integer.parseInt(st.nextToken());
			pq.offer(new Edge(A, B, C));
		}
		
		int cnt = 0;
		int ans = 0;
		
		while (cnt < V - 1) {
			
			Edge e = pq.poll();
			if (union(e.v1, e.v2)) {
				ans += e.weight;
				cnt++;
			}
		}
		
		System.out.println(ans);
	}

	private static boolean union(int x, int y) {
		
		int a = find(x);
		int b = find(y);
		if (a == b) return false;
		p[a] = b;
		
		return true;
	}

	private static int find(int x) {
		
		if (x == p[x]) return p[x];
		return p[x] = find(p[x]);
	}

}